stake-pool/program/src/processor.rs

```diff
@@ -19,65 +19,69 @@ use {
         decode_error::DecodeError,
         entrypoint::ProgramResult,
         msg,
-        native_token::sol_to_lamports,
+        native_token::LAMPORTS_PER_SOL,
         program::{invoke, invoke_signed},
         program_error::PrintProgramError,
         program_error::ProgramError,
         program_pack::Pack,
         pubkey::Pubkey,
         rent::Rent,
         stake_history::StakeHistory,
-        system_instruction,
+        system_instruction, system_program,
         sysvar::Sysvar,
     },
     spl_token::state::Mint,
 };
 
-/// Program state handler.
-pub struct Processor {}
-impl Processor {
-    /// Returns validator address for a particular stake account
-    fn get_validator(stake_account_info: &AccountInfo) -> Result<Pubkey, ProgramError> {
-        let stake_state: stake_program::StakeState = deserialize(&stake_account_info.data.borrow())
-            .or(Err(ProgramError::InvalidAccountData))?;
-        match stake_state {
-            stake_program::StakeState::Stake(_, stake) => Ok(stake.delegation.voter_pubkey),
-            _ => Err(StakePoolError::WrongStakeState.into()),
-        }
-    }
-
-    /// Checks if validator stake account is a proper program address
-    fn is_validator_stake_address(
-        vote_account: &Pubkey,
-        program_id: &Pubkey,
-        stake_pool_info: &AccountInfo,
-        stake_account_info: &AccountInfo,
-    ) -> bool {
-        // Check stake account address validity
-        let (stake_address, _) =
-            crate::find_stake_program_address(&program_id, &vote_account, &stake_pool_info.key);
-        stake_address == *stake_account_info.key
+const MAXIMUM_ACTIVE_STAKE: u64 = LAMPORTS_PER_SOL / 1_000;
```

Comment:
This constant seems named poor to me.  It probably could just be inlined at it's single use with a comment about why we can't use exactly 1 SOL

Reply:
But thinking about this more, could we still require exactly 1 SOL by splitting off any excess after warmup, so that the account to add has exactly 1 SOL?  
Alternatively, what if the stake account had the exact rent minimum instead of 1 SOL.  I don't think fractional rewards matter for the initial validator's stake account

Reply:
I can definitely inline it, no problem.

Otherwise, we can enforce the exact 1 SOL by doing a split.  Since there's a tolerance, you can already do that if you want!  It seemed like a headache for the staker to also remember a tiny stake account, since you'll have to do:

* split tiny stake
* deactivate
* wait 1 epoch to withdraw

The stake account can't have the exact rent minimum unfortunately -- all stake accounts must have at least 1 lamport above rent-exemption in order to be initialized.  Unless you imagine that it's an uninitalized stake account, in which case on the first deposit, we can split everything from the user's stake into the validator account.  This creates another risk of someone bricking your validator account -- if someone deposits 1 lamport, it becomes essentially impossible for anyone else to deposit since that requires a merge.

Reply:
Taking a step back, I think there's a different way.
1. There'll soon exist a fully inactive "reserve" stake account that holds the stake that is not currently allocated to a specific validator in the pool.
2. SOL from the "reserve" stake account could be used by the program to create and activate the initial stake account for the newly added validator.  The staker authority no longer needs to provide any of their SOL.

This means that initially when the pool is bootstrapped there'll need to be some initial SOL allocated to the reserve, but that's ok and the pool creator should get proper pool tokens back in return.  This seems to make pool bootstrapping pretty nice, logically:
The pool creator who has a pot of SOL creates a stake pool for it, with no validators and just a reserve stake account.  Creator then finds a stake authority that will start adding validators and managing to the pool, and (possibly) soliciting for other pool deposits.

Reply:
What about requiring a minimum balance of 1 SOL in all validator stake accounts?

Reply:
Just so I'm sure we're on the same page, the minimum requirement of 1 SOL is there, and this *also* adds the maximum check.  Since the staker isn't getting pool tokens in return, I added the maximum check to be sure they don't lose out more than needed, but I can certainly remove it.  Is that what you're suggesting?

Reply:
> I can't redeem my pool tokens (except by tiny bits as rewards are accrued)

I think I'm stuck on this point. With a 1 SOL minimum, a malicious staker could only create 10 validator stake accounts from my deposit. It's slightly annoying that I'd need to execute 10 transactions to recover my deposit but this doesn't seem too bad and the cli could wallpaper over this.  I'm probably just not getting *it* yet.

Reply:
The problem is that we maintain the 1 SOL minimum on withdrawals, so a validator account can't ever go under 1 SOL.  If we allow the accounts to go below that, then deposits will be broken for everyone after that.

So given that, how would I get my 10 SOL back if everything goes through the reserve account?

Reply:
What if
* Adding a new validator to the pool always requires the staker contribute the 1 SOL minimum (same for the reserve account at pool creation time).   The staker does not receive pool tokens for this 1 SOL minimum.
* Removing a validator from the pool requires that the staker draw that validator's stake balance back down to 1 SOL, at which point they can close the stake account and reclaim their 1 SOL (which did not receive any rewards).

Reply:
That's what this PR currently does :smile:

Reply:
The question seems to be this "maximum" amount, which I added as a courtesy so stakers are forbidden from putting too much SOL that they won't get pool tokens for.

Reply:
> That's what this PR currently does 😄

lol :shipit: 

Maybe an exact value instead of a maximum then?

Reply:
> lol :shipit: 

To me, it's a good sign that we got to the same answer :sunglasses: 

> Maybe an exact value instead of a maximum then?

We can do that, it means that every time you want to add a validator to the pool, you'll need to do

* create validator stake account
* wait until it's activated
* split just the rewards gained into another stake account
* add the validator stake account to the pool
* reclaim the rewards later

Those extra steps to split and reclaim for tiny rewards seemed annoying, but as the client, if you're ok with it, I'm down to do an exact value.

Reply:
How about this:
1. Deposit 1 SOL (without receiving pool tokens) into the reserve pool (always a fully inactive stake account)
2. Split 1 SOL from the reserve pool into a new stake account from the new validator and delegate it

Reply:
So essentially, we combine `CreateValidatorStakeAccount` and `AddValidatorToPool`, and the staker provides the funds?  That works for me.  In that case, do you mind if that's done in a later PR?  And I'll remove the maximum stake check in this one.

Reply:
> do you mind if that's done in a later PR

wfm